streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala

```diff
@@ -295,9 +297,42 @@ class KStream[K, V](val inner: KStreamJ[K, V]) {
    * @return a [[KStream]] that contains more or less records with new key and value (possibly of different type)
    * @see `org.apache.kafka.streams.kstream.KStream#transform`
    */
-  def transform[K1, V1](transformerSupplier: () => Transformer[K, V, KeyValue[K1, V1]],
+  def transform[K1, V1](transformerSupplier: TransformerSupplier[K, V, KeyValue[K1, V1]],
                         stateStoreNames: String*): KStream[K1, V1] =
-    inner.transform(transformerSupplier.asTransformerSupplier, stateStoreNames: _*)
+    inner.transform(transformerSupplier, stateStoreNames: _*)
+
+  /**
+    * Transform each record of the input stream into zero or more records in the output stream (both key and value type
+    * can be altered arbitrarily).
+    * A `Transformer` is applied to each input record and computes zero or more output records. In order to assign a
+    * state, the state must be created and registered beforehand via stores added via `addStateStore` or `addGlobalStore`
+    * before they can be connected to the `Transformer`
+    *
+    * @param transformer the `Transformer` instance
+    * @param stateStoreNames     the names of the state stores used by the processor
+    * @return a [[KStream]] that contains more or less records with new key and value (possibly of different type)
+    * @see `org.apache.kafka.streams.kstream.KStream#transform`
+    */
+  @deprecated("Use transform(TransformerSupplier,String) instead.", "2.0")
+  def transform[K1, V1](transformer: => Transformer[K, V, (K1, V1)], stateStoreNames: String*): KStream[K1, V1] = {
+    //noinspection ConvertExpressionToSAM // due to 2.11
+    val transformerSupplierJ: TransformerSupplier[K, V, KeyValue[K1, V1]] =
+      new TransformerSupplier[K, V, KeyValue[K1, V1]] {
+        override def get(): Transformer[K, V, KeyValue[K1, V1]] =
+          new Transformer[K, V, KeyValue[K1, V1]] {
+            override def transform(key: K, value: V): KeyValue[K1, V1] =
+              transformer.transform(key, value) match {
```

Comment:
We should add in the document that users should not use a transformer instance as the parameter.

Reply:
done.

Reply:
What do you think about just throwing an explanatory exception instead?